vm/src/stdlib/itertools.rs

```diff
@@ -488,6 +490,196 @@ mod decl {
         }
     }
 
+    struct GroupByState {
+        current_value: Option<PyObjectRef>,
+        current_key: Option<PyObjectRef>,
+        next_group: bool,
+        grouper: Option<Weak<PyObject<PyItertoolsGrouper>>>,
+    }
+
+    impl fmt::Debug for GroupByState {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            f.debug_struct("PyItertoolsGroupBy")
+                .field("current_value", &self.current_value)
+                .field("current_key", &self.current_key)
+                .field("next_group", &self.next_group)
+                .finish()
+        }
+    }
+
+    impl GroupByState {
+        fn is_current(&self, grouper: &PyItertoolsGrouperRef) -> bool {
+            if let Some(ref current_weak_grouper) = self.grouper {
+                if let Some(current_grouper) = current_weak_grouper.upgrade() {
+                    grouper.is(&current_grouper)
+                } else {
+                    false
+                }
+            } else {
+                false
+            }
+        }
+    }
+
+    #[pyclass(name = "groupby")]
+    struct PyItertoolsGroupBy {
+        iterable: PyObjectRef,
+        key_func: Option<PyObjectRef>,
+        state: Mutex<GroupByState>,
```

Comment:
1. Is Mutex prefered to RwLock here? (I didn't check details well, so it could be a stupid question)
2. I thought about state design that checking initial state by optional `current_key` is a good idea or not. `Option<GroupByState>` probably express empty initial state the best, but it seems it is a duplicated option wrapping for most of use case. I want to listen your opinion about those designs.

Reply:
After thinking about it more, there are actually 3 or 4 different states that it can be in so maybe the following is a better way to model it:
```rust
enum GroupByState {
    Initial,
    InGroup {
        key: PyObjectRef,
        value: Option<PyObjectRef>,
        grouper: Option<Weak<PyObject<PyItertoolsGrouper>>>,
    },
    NextGroup {
        next_key: PyObjectRef,
        next_value: PyObjectRef,
    },
}
```